File: frontend/src/utils/scoring.ts
/**
 * Competition AI Code Submission Scoring Utility.
 *
 * Mathematically calculates the overall AI Evaluation Score as the unweighted
 * arithmetic mean of all 6 competition criteria:
 * 1. Code Quality
 * 2. Security
 * 3. Efficiency
 * 4. Testing
 * 5. Accessibility
 * 6. Problem Statement Alignment
 *
 * Formula:
 *   overallScore = (codeQuality + security + efficiency + testing + accessibility + problemAlignment) / 6
 * Rounded to 2 decimal places.
 */

export interface CriterionDetail {
  score: number;
  weight?: number;
  label?: string;
  evidence?: string;
  measured_type?: string;
}

export interface CriteriaMap {
  code_quality?: CriterionDetail | number;
  security?: CriterionDetail | number;
  efficiency?: CriterionDetail | number;
  testing?: CriterionDetail | number;
  accessibility?: CriterionDetail | number;
  problem_alignment?: CriterionDetail | number;
  problem_statement_alignment?: CriterionDetail | number;
  [key: string]: any;
}

export function calculateOverallScore(criteria?: CriteriaMap | null): number | null {
  if (!criteria || typeof criteria !== 'object') {
    return null;
  }

  const keys = [
    'code_quality',
    'security',
    'efficiency',
    'testing',
    'accessibility',
    'problem_alignment'
  ];

  const scores: number[] = [];

  for (const key of keys) {
    let item = criteria[key];
    if ((item === undefined || item === null) && key === 'problem_alignment') {
      item = criteria['problem_statement_alignment'];
    }

    if (item === undefined || item === null) {
      console.warn(`[Scoring] Missing criterion: '${key}'`);
      return null;
    }

    const val = typeof item === 'number' ? item : item.score;
    if (val === undefined || val === null || typeof val !== 'number' || Number.isNaN(val)) {
      console.warn(`[Scoring] Invalid score for criterion '${key}':`, val);
      return null;
    }

    if (val < 0 || val > 100) {
      console.warn(`[Scoring] Criterion score out of range (0-100) for '${key}':`, val);
      return null;
    }

    scores.push(val);
  }

  if (scores.length !== 6) {
    return null;
  }

  const sum = scores.reduce((acc, curr) => acc + curr, 0);
  const avg = sum / 6;
  return Math.round(avg * 100) / 100;
}

File: backend/app/analysis/scoring.py
"""
AI Evaluation Scoring Utility Module
Implements equal weighting (16.6667% per criterion) for the 6 PromptWars Competition Criteria:
1. Code Quality
2. Security
3. Efficiency
4. Testing
5. Accessibility
6. Problem Statement Alignment

Formula:
  Overall Score = (code_quality + security + efficiency + testing + accessibility + problem_alignment) / 6
"""

from typing import Dict, Any, Union

def calculate_overall_score(
    code_quality: float,
    security: float,
    efficiency: float,
    testing: float,
    accessibility: float,
    problem_alignment: float
) -> float:
    """Calculates transparent overall score as exact arithmetic mean of 6 criteria."""
    total = (
        float(code_quality) +
        float(security) +
        float(efficiency) +
        float(testing) +
        float(accessibility) +
        float(problem_alignment)
    )
    return round(total / 6.0, 2)


def calculate_overall_from_criteria_dict(criteria: Dict[str, Any]) -> float:
    """Calculates overall score from a criteria dictionary containing the 6 keys."""
    cq = criteria.get("code_quality", {}).get("score", criteria.get("code_quality", 0))
    sec = criteria.get("security", {}).get("score", criteria.get("security", 0))
    eff = criteria.get("efficiency", {}).get("score", criteria.get("efficiency", 0))
    tst = criteria.get("testing", {}).get("score", criteria.get("testing", 0))
    a11y = criteria.get("accessibility", {}).get("score", criteria.get("accessibility", 0))
    psa = criteria.get("problem_alignment", {}).get("score", criteria.get("problem_alignment", 0))

    if isinstance(cq, dict): cq = cq.get("score", 0)
    if isinstance(sec, dict): sec = sec.get("score", 0)
    if isinstance(eff, dict): eff = eff.get("score", 0)
    if isinstance(tst, dict): tst = tst.get("score", 0)
    if isinstance(a11y, dict): a11y = a11y.get("score", 0)
    if isinstance(psa, dict): psa = psa.get("score", 0)

    return calculate_overall_score(cq, sec, eff, tst, a11y, psa)
